### ISSUE TITLE
Discovery: moment.js alternative

### ISSUE THREAD

**Issue by rianfowler** [department-of-veterans-affairs/va.gov-team#6763]
@markgreenburg commented on [Fri Aug 03 2018](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/11963)

## Problem:
Moment.js is ~10x the size of React, gzipped (>60K). This seems to be the case despite removing locales via `new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/)`. Further, imports can't be modularized so we can't take advantage of tree shaking. This is a really big dependency, and we don't really use many fancy features to make moment worth all this weight.

## Discovery:
- [ ] Estimate & quantify benefit from switching to an alternative library
- [ ] Document all the different `moment` functions we use in vets-website
- [ ] Explore alternatives
  - Two mentioned [here](https://github.com/moment/moment/issues/2373) are [date-fns](https://github.com/date-fns/date-fns) and [day.js](https://github.com/iamkun/dayjs)
- [ ] Explore lift to switch. Note that there're also a `moment-setup.js` in `platforms/site-wide` as well as a `utils/date/index.js` in addition to the usage across our apps.

---

@cvalarida commented on [Fri Aug 03 2018](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/11963#issuecomment-410384512)

@markgreenburg Did you measure this in the bundle served to the client or using VSCode's fancy new import size feature? I'm wondering if the latter doesn't factor in that webpack ignore plugin.

**Comment by bkjohnson** [comment]
Here are the `moment` functions that are currently in use:

- `.add()`
- `.asDays()`
- `.clone()`
- `.daysInMonth()`
- `.diff()`
- `.duration()`
- `.endOf()`
- `.format()`
- `.fromNow()`
- `.isDST()` - no 1:1 equivalent in `date-fns`
- `.isAfter()`
- `.isBefore()`
- `.isBetween()`
- `.isSame()`
- `.isSameOrBefore()`
- `.isSameOrAfter()`
- `.isValid()`
- `.locale()`
- `.month()`
- `.parseZone()`
- `.startOf()`
- `.subtract()`
- `.toDate()`
- `.toISOString()`
- `.unix()`
- `updateLocale()` - in `moment-setup.js`
- `.utc()` - no 1:1 equivalent in `date-fns`
- `.utcOffset()`- no 1:1 equivalent in `date-fns`


In order to make the switch to `date-fns` we would have to refactor platform/app code and tests.  We may encounter code with little or no test coverage that is being modified, and we would have to add tests in those scenarios to ensure that the dependency switch & associated refactoring doesn't break anything.

**Comment by bkjohnson** [comment]
I think it may be wise to hold off on this for a bit.  `date-fns` [doesn't provide a way of changing the default locale](https://github.com/date-fns/date-fns/issues/816), and we've run into a few [formatting issues](https://github.com/department-of-veterans-affairs/vets-website/pull/12112#discussion_r403391690) where the default doesn't match what our content guidelines demand.  With `moment`, we circumvented those issues in the `moment-setup.js` file, but `date-fns` doesn't give us a comparable `updateLocale()` function.

We _could_ pass an `options` object each time we called `format`, or we could create a wrapper function that used the locale that we wanted, but those approaches are just extra hassle that would require more maintenance.
